plugins/catalog-backend/src/processing/TaskPipeline.ts

```
@@ -72,50 +74,58 @@ export function startTaskPipeline<T>(options: Options<T>) {
+      if (state.inFlightCount <= lowWatermark) {
+        const loadCount = highWatermark - state.inFlightCount;
+        const loadedItems = await Promise.resolve()
+          .then(() => loadTasks(loadCount))
+          .catch(() => {
+            // Silently swallow errors and go back to sleep to try again; we
+            // delegate to the loadTasks function itself to catch errors and log
+            // if it so desires
+            return [];
+          });
+        if (loadedItems.length && !abortSignal.aborted) {
+          state.inFlightCount += loadedItems.length;
+          for (const item of loadedItems) {
+            Promise.resolve()
+              .then(() => processTask(item))
+              .catch(() => {
+                // Silently swallow errors and go back to sleep to try again; we
+                // delegate to the processTask function itself to catch errors
+                // and log if it so desires
+              })
+              .finally(() => {
+                state.inFlightCount -= 1;
+                barrier.release();
+              });
+          }
         }
+      }
 
-        // For each item we complete we check if it's time to load more
-        inFlightCount -= 1;
-        maybeLoadMore();
-      });
-    });
-
-    // We might have processed some tasks while we where loading, so check if we can load more
-    if (loadedItems.length > 1) {
-      maybeLoadMore();
+      await barrier.wait();
     }
   }
 
-  // This interval makes sure that we load in new items if the loop runs
-  // dry because of the lack of available tasks. As long as there are
-  // enough items to process this will be a noop.
-  const intervalId = setInterval(() => {
-    maybeLoadMore();
-  }, pollingIntervalMs);
+  pipelineLoop().catch(_error => {
+    // This should be impossible, but if it did happen, it would signal a
+    // programming error inside the loop (errors should definitely be caught
+    // inside of it).
+  });
```

We don't want this to fail silently do we though? Probably best to re-throw as an uncaught exception to crash the backend?

---

We don't crash on those anymore do we? 🤔

---

Alright, updated this with a rethrow, it'll make it log which is excellent either way.